### PR TITLE
Use aws-sdk-go to communicate with AWS

### LINKED ISSUE #2: Error doing multipart upload
I am getting the following error trying to upload a large (~1GB) box:

```
==> virtualbox-iso: Running post-processor: vagrant-s3
==> virtualbox-iso (vagrant-s3): Preparing to upload box for 'virtualbox' provider to S3 bucket 'vagrant.lmars.net'
    virtualbox-iso (vagrant-s3): Box to upload: packer_virtualbox-iso_virtualbox.box (1060157666 bytes)
    virtualbox-iso (vagrant-s3): Fetching latest manifest
    virtualbox-iso (vagrant-s3): Generating checksum
    virtualbox-iso (vagrant-s3): Checksum is fe24c05d9aac09a1d2f81965a3c25e01519192091c6f233a3f160a0a898dae48
    virtualbox-iso (vagrant-s3): Adding virtualbox 20141008.0 box to manifest
    virtualbox-iso (vagrant-s3): Uploading box to S3: vagrant/boxes/20141008.0/packer_virtualbox-iso_virtualbox.box
    virtualbox-iso (vagrant-s3): File size > 100MB. Initiating multipart upload
Build 'virtualbox-iso' errored: 1 error(s) occurred:

* Post-processor failed: Your socket connection to the server was not read from or written to within the timeout period. Idle connections will be closed.
```

### LINKED ISSUE #25: Use aws-sdk-go instead of goamz/s3
goamz/s3 seems to be deprecated and currently doesn't support some S3 regions at all (e.g. eu-central-1). Currently if you try to set region to eu-central-1 you get this error:

```
* Post-processor failed: The authorization mechanism you have provided is not supported. Please use AWS4-HMAC-SHA256
```

References:
https://github.com/mitchellh/goamz/issues/166
https://github.com/mitchellh/goamz/pull/230

### PR DESCRIPTION
**Pull request by Jalle19**
Closes #2, closes #25.

There are several reasons for this:

- it's officially supported and actively maintained
- streaming uploads of large files is included in the library which leads to a drastic reduction in complexity for our use case. The upload process is also much more robust, it even handles putting the computer to sleep and waking it up afterwards.

I based this branch on my automatic versioning branch so it will need rebasing before it's good to go.

@lmars can you review? The logic in the main method was starting to get a bit long so I factored out some parts to separate functions.

### CODE REVIEW

**Comment by Jalle19** [comment]
Forgot to say that the credential validation logic from the `Configure` step is still missing.

**Comment by Jalle19** [comment]
@lmars I re-introduced the validation logic, it now attempts to get information about the target bucket as a check:

```
Errors validating build 'virtualbox-ovf'. 1 error(s) occurred:

* Unable to access the bucket eu-west-1, make sure your credentials are valid and have sufficient permissions

Errors validating build 'vmware-vmx'. 1 error(s) occurred:

* Unable to access the bucket eu-west-1, make sure your credentials are valid and have sufficient permissions
```

File: post-processor.go [hunk 1]
```diff
@@ -268,14 +268,56 @@ func (p *PostProcessor) putManifest(manifest *Manifest) error {
...
+	// open the file
+	file, err := os.Open(boxPath)
+	if err != nil {
+		return -1, err
+	}
+
+	defer file.Close()
+
+	// return the file size
+	info, err := file.Stat()
```
Comment (lmars) [inline post-processor.go]: No need to open the file first, just use [os.Stat](https://golang.org/pkg/os/#Stat).

**Comment by lmars** [comment]
@Jalle19 please rebase on master then I'll take another look.

**Comment by Jalle19** [comment]
@lmars rebased and dropped `func getFileSize()` in favor of using `os.Stat()` directly.

**Comment by Jalle19** [comment]
For the record I haven't tested the generation of signed URLs yet, I'll try to do it today.

**Comment by lmars** [comment]
Thanks!

**Pull request merged**